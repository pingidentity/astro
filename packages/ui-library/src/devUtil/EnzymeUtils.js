import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
