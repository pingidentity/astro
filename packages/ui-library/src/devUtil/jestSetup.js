
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
