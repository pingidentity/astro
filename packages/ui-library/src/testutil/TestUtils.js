/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
