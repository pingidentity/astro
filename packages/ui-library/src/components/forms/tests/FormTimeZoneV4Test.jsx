import React from "react";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import ReactTestUtils from "react-dom/test-utils";
import ReactDOM from "react-dom";
import TestUtils from "../../../testutil/TestUtils";
import FormTimeZone from "../FormTimeZone";
import KeyboardUtils from "../../../util/KeyboardUtils.js";
import moment from "moment-timezone";
import momentMetadata from "moment-timezone/data/meta/latest.json";
import _ from "underscore";

jest.mock("popper.js");
jest.mock("react-portal");

// This file is a copy of the standard Time Zone tests with all the v4 flags set by default
describe("FormTimeZone v4", function () {
    const componentId = "timezone";
    const initialValue = "America/Denver";
    const countryLabel="select a Country";

    function prepCountryMetaData () {
        var countryMetadata = [];
        for (var countryCode in momentMetadata.countries) {
            countryMetadata.push(momentMetadata.countries[countryCode]);
        }
        countryMetadata = _.sortBy(countryMetadata, function (country) {
            return country.name;
        });

        return countryMetadata;
    }
    var countryMetadata = prepCountryMetaData();

    function prepZoneMetaData () {
        var currentUtcTime = moment().utc(),
            zoneNames = moment.tz.names();

        return zoneNames.map(function (zone) {
            return {
                abbr: moment.tz(zone).zoneAbbr(),
                name: zone,
                offset: moment.tz(currentUtcTime, zone).format("Z")
            };
        });
    }
    var zoneMetadata = prepZoneMetaData();

    function getComponent (opts) {
        const options = _.defaults(opts || {}, {
            countryLabel: countryLabel,
            "data-id": componentId,
            initialState: {
                value: initialValue,
            },
            flags: [ "use-portal", "p-stateful" ],
        });
        return ReactTestUtils.renderIntoDocument(<FormTimeZone {...options} />);
    }

    function getValueLink (component) {
        return TestUtils.findRenderedDOMNodeWithDataId(component, componentId + "-collapsible-link");
    }

    function getSearchInput (component) {
        return TestUtils.findRenderedDOMNodeWithDataId(component, "searchBox-input");
    }

    function getRows (component) {
        return ReactTestUtils.scryRenderedDOMComponentsWithClass(component, "button-menu__button");
    }

    function getMenu (component) {
        return TestUtils.findRenderedDOMNodeWithClass(component, "button-menu__scroller");
    }

    function getCountryZones (countryAbbr) {
        const countryZoneNames = momentMetadata.countries[countryAbbr].zones;
        let countryZones;

        countryZones = zoneMetadata.filter(function (zone) {
            return countryZoneNames.indexOf(zone.name) > -1;
        });
        countryZones = _.sortBy(countryZones, function (country) {
            return country.offset;
        });

        return countryZones;
    }

    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();

    beforeEach(function () {
        window.addEventListener.mockClear();
        window.removeEventListener.mockClear();
    });

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <FormTimeZone
                flags={[ "use-portal", "p-stateful" ]}
                showClear
                value="25102019"
            />
        );
    });

    it("renders the value link", function () {
        const component = getComponent();
        const valueLink = getValueLink(component);

        expect(valueLink).toBeTruthy();
        expect(valueLink.textContent).toEqual(initialValue);
    });

    it("stateful: renders the value link without underscores", function () {
        const component = getComponent({ value: "America/Los_Angeles" });
        const valueLink = getValueLink(component);

        expect(valueLink).toBeTruthy();
        expect(valueLink.textContent).toEqual("America/Los Angeles");
    });

    it("stateless: using getZoneNameDisplayValue renders value link without underscores", function () {
        const component = getComponent({ value: FormTimeZone.getZoneNameDisplayValue("America/Fort_Nelson") });
        const valueLink = getValueLink(component);

        expect(valueLink).toBeTruthy();
        expect(valueLink.textContent).toEqual("America/Fort Nelson");
    });

    it("renders the menu when the link is clicked", function () {
        const component = getComponent();
        const valueLink = getValueLink(component);
        let menu;

        menu = getMenu(component);
        expect(menu).toBeFalsy();

        ReactTestUtils.Simulate.click(valueLink);

        menu = getMenu(component);
        expect(menu).toBeTruthy();
    });

    it("renders the countries with the proper text", function () {
        const component = getComponent({ open: true });
        const countryRows = getRows(component);

        expect(countryRows.length).toEqual(countryMetadata.length);
        expect(countryRows[0].textContent).toEqual(countryMetadata[0].name);
        expect(countryRows[countryRows.length - 1].textContent)
            .toEqual(countryMetadata[countryMetadata.length - 1].name);
    });

    it("filters the country list when search text is entered", function () {
        const searchString = "United";
        const component = getComponent({
            open: true,
            searchString: searchString
        });
        const searchInput = getSearchInput(component);
        const countryRows = getRows(component);

        expect(searchInput.value).toEqual(searchString);
        expect(countryRows.length).toEqual(2); // United Arab Emirates & US
        expect(countryRows[0].textContent).toEqual(momentMetadata.countries.AE.name);
        expect(countryRows[1].textContent).toEqual(momentMetadata.countries.US.name);
    });

    it("clears the search text when the clear search-text link is clicked", function () {
        const searchString = "United";
        const component = getComponent({
            open: true,
            initialState: { searchString }
        });
        const searchInput = getSearchInput(component);
        const clearSearch = TestUtils.findRenderedDOMNodeWithClass(component, "clear-search");

        expect(searchInput.value).toEqual(searchString);
        ReactTestUtils.Simulate.click(clearSearch);
        expect(searchInput.value).toEqual("");
    });

    it("selects the country when the country is clicked", function () {
        const component = getComponent({ open: true });
        const countryRows = getRows(component);
        let selectedCountry;

        ReactTestUtils.Simulate.click(countryRows[3]);
        selectedCountry = TestUtils.findRenderedDOMNodeWithDataId(component, "selected-country");
        expect(selectedCountry.textContent).toContain(countryMetadata[3].name);
    });

    it("selects the top country when the ENTER key is pressed while searching", function () {
        const component = getComponent({ open: true });
        const countryRows = getRows(component);
        const searchInput = getSearchInput(component);
        let selectedCountry;

        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ENTER });
        selectedCountry = TestUtils.findRenderedDOMNodeWithDataId(component, "selected-country");
        expect(selectedCountry.textContent).toContain(countryMetadata[0].name);
        expect(countryRows[0].className).toContain("selected");
    });

    it("selects the second country when the UP/DOWN-ARROW keys are pressed then ENTER key is pressed", function () {
        const component = getComponent({ open: true });
        const countryRows = getRows(component);
        const searchInput = getSearchInput(component);
        let selectedCountry;

        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ARROW_DOWN });
        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ARROW_DOWN });
        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ARROW_UP });
        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ENTER });
        selectedCountry = TestUtils.findRenderedDOMNodeWithDataId(component, "selected-country");
        expect(selectedCountry.textContent).toContain(countryMetadata[1].name);
        expect(countryRows[1].className).toContain("selected");
    });

    it("renders the zones with the proper text", function () {
        const countryAbbr = "US";
        const component = getComponent({
            open: true,
            filterByCountry: countryAbbr
        });
        const zoneRows = getRows(component);
        const dataZones = getCountryZones(countryAbbr);
        let zoneName;

        expect(zoneRows.length).toEqual(dataZones.length);

        /* loop through the rendered zones and make sure each one can be found in the zones pulled from the metadata
           for the selected country */
        for (var i=0; i<zoneRows.length; i+=1) {
            zoneName = TestUtils.findRenderedDOMNodeWithClass(zoneRows[i], "timezone-name").textContent;
            expect(zoneName).toEqual(FormTimeZone.getZoneNameDisplayValue(dataZones[i].name));
        }
    });

    it("renders no zones if a bad country code is passed in", function () {
        const countryAbbr = "ALL";
        const component = getComponent({
            open: true,
            filterByCountry: countryAbbr
        });
        const zoneRows = getRows(component);

        expect(zoneRows.length).toEqual(0);
    });

    it("selects the zone when the zone is clicked", function () {
        const countryAbbr = "US";
        const callback = jest.fn();
        const component = getComponent({
            open: true,
            filterByCountry: countryAbbr,
            onZoneChange: callback,
        });
        const testIndex = 4;
        const zoneRows = getRows(component);
        const zoneRowName = TestUtils.findRenderedDOMNodeWithClass(zoneRows[testIndex], "timezone-name").textContent;
        let valueLink;

        ReactTestUtils.Simulate.click(zoneRows[testIndex]);
        valueLink = getValueLink(component);
        expect(valueLink.textContent).toEqual(zoneRowName);
    });

    it("selects the top zone when the ENTER key is pressed while searching", function () {
        const countryAbbr = "US";
        const component = getComponent({
            open: true,
            filterByCountry: countryAbbr
        });
        const searchInput = getSearchInput(component);
        const testIndex = 0;
        const zoneRows = getRows(component);
        const zoneRowName = TestUtils.findRenderedDOMNodeWithClass(zoneRows[testIndex], "timezone-name").textContent;
        let valueLink;

        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ENTER });
        valueLink = getValueLink(component);
        expect(valueLink.textContent).toEqual(zoneRowName);
        expect(zoneRows[testIndex].className).toContain("selected");
    });

    it("selects the second zone when the DOWN-ARROW key is pressed then ENTER key is pressed", function () {
        const callback = jest.fn();
        const countryAbbr = "US";
        const component = getComponent({
            open: true,
            filterByCountry: countryAbbr,
            onZoneChange: callback,
        });
        const searchInput = getSearchInput(component);
        const zoneRows = getRows(component);
        const zoneRowName = TestUtils.findRenderedDOMNodeWithClass(zoneRows[1], "timezone-name").textContent;
        let valueLink;

        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ARROW_DOWN });
        expect(zoneRows[1].className).toContain("selected");
        ReactTestUtils.Simulate.keyDown(searchInput, { keyCode: KeyboardUtils.KeyCodes.ENTER });
        valueLink = getValueLink(component);
        expect(valueLink.textContent).toEqual(zoneRowName);
    });

    it("calls global click handler when a click occurs outside of component", function () {
        const component = getComponent({
            stateless: true,
            open: true,
            onSearch: jest.fn(),
            onToggle: jest.fn(),
            onValueChange: jest.fn()
        });

        global.getSelection = jest.fn();
        global.getSelection.mockReturnValue({
            toString: () => "",
        });

        const stateless = ReactTestUtils.findRenderedComponentWithType(component, FormTimeZone._statelessComponent);
        const handler = stateless._onGlobalClick;

        expect(component.props.onToggle).not.toBeCalled();

        // click outside
        handler({ target: document.body });

        // hack, for now, to get this test to work
        // @TODO : remove the following line
        ReactTestUtils.Simulate.click(getValueLink(component));

        expect(component.props.onToggle).toBeCalled();
    });

    it("skips the global click handler if not open and click on component", function () {
        const component = getComponent({
            stateless: true,
            onToggle: jest.fn(),
            onSearch: jest.fn(),
            onValueChange: jest.fn()
        });
        const handler = TestUtils.findMockCall(window.addEventListener, "click")[1];

        // click on component
        handler({ target: { dataset: component } });

        expect(component.props.onToggle).not.toBeCalled();
    });

    it("detaches global listeners on unmount", function () {
        const component = getComponent({
            stateless: true,
            onToggle: jest.fn(),
            onSearch: jest.fn(),
            onValueChange: jest.fn()
        });
        const stateless = ReactTestUtils.findRenderedComponentWithType(component, FormTimeZone._statelessComponent);

        expect(window.addEventListener).toBeCalledWith("click", stateless._onGlobalClick);

        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(component).parentNode);
        expect(window.removeEventListener).toBeCalledWith("click", stateless._onGlobalClick);
    });

    it("calls function for list placement on componentDidUpdate", function () {
        const component = getComponent({
            open: true,
            stateless: true
        });
        const stateless = ReactTestUtils.findRenderedComponentWithType(component, FormTimeZone._statelessComponent);

        stateless._setListPosition = jest.fn();
        stateless.componentDidUpdate();
        expect(stateless._setListPosition).toBeCalled();
    });

    it("does not render the error when not provided", function () {
        const component = getComponent();
        const errorIcon = TestUtils.findRenderedDOMNodeWithDataId(component, componentId + "-error-message-icon");
        const errorMessage = TestUtils.findRenderedDOMNodeWithDataId(component, componentId + "-error-message");

        expect(errorIcon).toBeFalsy();
        expect(errorMessage).toBeFalsy();
    });

    it("renders the error message when provided", function () {
        const errorMessageText = "warning!";
        const component = getComponent({
            errorMessage: errorMessageText
        });
        const errorIcon = TestUtils.findRenderedDOMNodeWithDataId(component, "timezone-error-message-icon");
        const errorMessage = TestUtils.findRenderedDOMNodeWithDataId(component, "timezone-error-message");

        expect(errorIcon).toBeTruthy();
        expect(errorMessage).toBeTruthy();
        expect(errorMessage.textContent).toEqual(errorMessageText);
    });

    it("public function isValidTimeZone works properly", function () {
        const component = getComponent();
        const realZoneName = zoneMetadata[10].name;
        let checkZone = component.isValidTimeZone("badZoneName");

        expect(checkZone).toEqual(false);

        checkZone = component.isValidTimeZone(realZoneName);
        expect(checkZone).toBeTruthy();
        expect(checkZone.name).toEqual(realZoneName);
    });

    it("renders the displayValue when provided", function () {
        const displayValue = "MDT";
        const component = getComponent({
            open: true,
            value: "America/Denver",
            displayValue: displayValue
        });
        const linkValue = getValueLink(component);

        expect(linkValue.textContent).toEqual(displayValue);
    });

    it("does not render the label text when not specified", function () {
        const component = getComponent();
        const renderedLabelText = TestUtils.findRenderedDOMNodeWithDataId(component, "label");

        expect(renderedLabelText).toBeFalsy();
    });

    it("fires cannonball warning when p-stateful flag is not set", function() {
        console.warn = jest.fn();

        expect(console.warn).not.toBeCalled();
        getComponent({ flags: [ "use-portal" ] });
        expect(console.warn).toBeCalled();
    });

    it("doesn't fire cannonball warning when p-stateful flag is set", function() {
        console.warn = jest.fn();

        expect(console.warn).not.toBeCalled();
        getComponent({ flags: [ "use-portal", "p-stateful" ] });
        expect(console.warn).not.toBeCalled();
    });

    /*
    TODO: fix this test
    NOTE: Currently we've not been able to test the react-tooltip helpHint in it's shown state. Since the React Tooltip
    does not populate its content until it is displayed, the content of the tooltip cannot yet be tested.
    */
    it("renders the label text and help text when specified", function () {
        const labelText = "My TZ Picker";
        const labelHelpText = "Help text here!";
        const component = getComponent({
            labelText: labelText,
            labelHelpText: labelHelpText
        });
        const renderedHelpIcon = TestUtils.findRenderedDOMNodeWithClass(component, "icon-help");
        // const renderedHelpText = TestUtils.findRenderedDOMNodeWithClass(component, "tooltip-text");
        const renderedLabelText = TestUtils.findRenderedDOMNodeWithDataId(component, "label");

        expect(renderedLabelText.textContent).toContain(labelText);
        expect(renderedHelpIcon).toBeTruthy();
        // expect(renderedHelpText.textContent).toEqual(labelHelpText);
    });

});
