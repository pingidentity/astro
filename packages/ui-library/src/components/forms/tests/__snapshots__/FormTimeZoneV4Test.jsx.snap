// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTimeZone v4 data-id's don't change 1`] = `
Array [
  "formLabel",
  "tooltip-menu",
  "tooltip-menu-trigger",
  "undefined-collapsible-link",
]
`;
